src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1582 @@
+    if (provider == NULL) {
+        assert(0);
+        return;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    utils_mutex_destroy_not_free(&coarse_provider->lock);
+
+    ravl_foreach(coarse_provider->all_blocks, coarse_ravl_cb_rm_all_blocks_node,
+                 coarse_provider);
+    assert(coarse_provider->used_size == 0);
+
+    ravl_foreach(coarse_provider->upstream_blocks,
+                 coarse_ravl_cb_rm_upstream_blocks_node, coarse_provider);
+    assert(coarse_provider->alloc_size == 0);
+
+    ravl_delete(coarse_provider->upstream_blocks);
+    ravl_delete(coarse_provider->all_blocks);
+    ravl_delete(coarse_provider->free_blocks);
+
+    umf_ba_global_free(coarse_provider->name);
+
+    umf_ba_global_free(coarse_provider);
+}
+
+static umf_result_t
+create_aligned_block(coarse_memory_provider_t *coarse_provider,
+                     size_t orig_size, size_t alignment, block_t **current) {
+    (void)orig_size; // unused in the Release version
+    int rv;
+
+    block_t *curr = *current;
+
+    // In case of non-zero alignment create an aligned block what would be further used.
+    uintptr_t orig_data = (uintptr_t)curr->data;
+    uintptr_t aligned_data = ALIGN_UP(orig_data, alignment);
+    size_t padding = aligned_data - orig_data;
+    if (alignment > 0 && padding > 0) {
+        block_t *aligned_block = coarse_ravl_add_new(
+            coarse_provider->all_blocks, curr->data + padding,
+            curr->size - padding, NULL);
+        if (aligned_block == NULL) {
+            return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        }
+
+        curr->used = false;
+        curr->size = padding;
+
+        rv = free_blocks_add(coarse_provider->free_blocks, curr);
+        if (rv) {
+            return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        }
+
+        // use aligned block
+        *current = aligned_block;
+        assert((*current)->size >= orig_size);
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t
+split_rest_and_merge_new_block(coarse_memory_provider_t *coarse_provider,
+                               block_t *curr, size_t size) {
+    int rv;
+
+    // Split the rest of block and put the new block after the one that we use.
+    void *data = curr->data + size;
+
+    ravl_node_t *new_node = NULL;
+
+    block_t *new_block = coarse_ravl_add_new(coarse_provider->all_blocks, data,
+                                             curr->size - size, &new_node);
+    if (new_block == NULL) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    new_block->used = false;
+
+    // Try to merge new empty block with the next one.
+    new_node = free_block_merge_with_next(coarse_provider, new_node);
+    rv =
+        free_blocks_add(coarse_provider->free_blocks, get_node_block(new_node));
+    if (rv) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
```

Comment:
This is allways false. We are spiting free block. There cannot be a mergable free block next to it, as it will be already merged with block we are spiting (we even have debug check that ensures that - debug_verify_all_blocks_cb)

ps. if you remove it, please rename function too.

Reply:
Done